### PR TITLE
[Tensor] add module handler for linear

### PR DESCRIPTION
**Pull request by Wesley-Jzy**
1. add module spec function
2. add ColoLinear as a demo
3. add unit test

### CODE REVIEW

File: colossalai/tensor/module_handlers/linear.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
+from .colo_module_handler import ColoModuleHandler
+from colossalai.tensor import ComputePattern, distspec
+from colossalai.core import global_context as gpc
+from colossalai.context.parallel_mode import ParallelMode
+
+class ColoLinear(ColoModuleHandler):
+    def __init__(self):
+        super(ColoLinear, self).__init__()
+        self._register_shard_params(['weight', 'bias'])
+        self._set_TP1D()
```
Comment (feifeibear) [inline colossalai/tensor/module_handlers/linear.py:10]: why set TP1D? ColoLinear can not use DP?

File: colossalai/tensor/module_utils.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
+from typing import Dict
+from colossalai.tensor import ColoParameter, ParallelAction, TensorSpec
+import torch
+
+_COLOSSAL_MODULES: Dict[str, object] = {}
+
+
+def register_colo_module(module: torch.nn.Module, colo_module):
```
Comment (feifeibear) [inline colossalai/tensor/module_utils.py:8]: type hinting is wrong!
the 1st arg should be Type[torch.nn.Module] but not a module instance?

File: colossalai/tensor/module_handlers/linear.py (hunk 1 above)
Comment (Wesley-Jzy) [reply colossalai/tensor/module_handlers/linear.py:10]: DP is a default pattern now. It's not in ComputePattern.

File: colossalai/tensor/module_utils.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+                    if cur_compute_pattern != compute_pattern:
+                        print(f'Invalid ColoParameter spec: Params in {module} have different compute_pattern.')
+                        raise
+            else:
+                continue
+            
+        if compute_pattern is not None:
+            if not handler.has_compute_pattern(compute_pattern):
+                print(f'Invalid ColoParameter spec: ComputePattern {compute_pattern} in {module} is not allowed.')
+                raise
```
Comment (feifeibear) [inline colossalai/tensor/module_utils.py:47]: raise an Exception!

File: colossalai/tensor/module_utils.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+def check_colo_module(module: torch.nn.Module, recursive=True):
+    if is_colo_module(module):
+        handler = get_colo_module(module)
+        param_names = handler.get_param_names()
+        compute_pattern = None
+        for param_name in param_names:
+            param = module.get_parameter(param_name)
+            if not isinstance(param, ColoParameter):
+                print(f'Invalid ColoParameter spec: {param} in {module} is not a ColoParameter .')
```
Comment (feifeibear) [inline colossalai/tensor/module_utils.py:31]: move the error message to Exception

File: colossalai/tensor/module_utils.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+                raise
+            if param.has_spec():  
+                cur_compute_pattern = param.spec.parallel_action.compute_pattern
+                if compute_pattern is None:
+                    compute_pattern = cur_compute_pattern
+                else:
+                    if cur_compute_pattern != compute_pattern:
+                        print(f'Invalid ColoParameter spec: Params in {module} have different compute_pattern.')
```
Comment (feifeibear) [inline colossalai/tensor/module_utils.py:39]: move the error message to Exception

File: colossalai/tensor/module_utils.py (hunk 2 above)
Comment (feifeibear) [reply colossalai/tensor/module_utils.py:8]: Also what is the type of colo_module?

File: colossalai/tensor/module_utils.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    if recursive == True:
+        for submodule in module.children():
+            check_colo_module(submodule, recursive=True)
+
+def init_colo_module(module: torch.nn.Module, parallel_action: ParallelAction, recursive=True, label='default'):
+    compute_pattern = parallel_action.compute_pattern
+    if is_colo_module(module):
+        # for each param
+        # set DistSpec and ParallelAction
+        handler = get_colo_module(module)
```
Comment (feifeibear) [inline colossalai/tensor/module_utils.py:79]: I expect get_colo_module returns a module. What do you mean a handler?

File: colossalai/tensor/module_utils.py (hunk 2 above)
Comment (Wesley-Jzy) [reply colossalai/tensor/module_utils.py:8]: Right. I will fix it.

File: colossalai/tensor/module_utils.py (hunk 6 above)
Comment (feifeibear) [reply colossalai/tensor/module_utils.py:79]: Please make the function's name explains itself.

File: tests/test_tensor/test_module_spec.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        bias.set_spec(spec)
+
+
+def run_with_spec(spec_init_func):
+    with ColoInitContext(device=get_current_device()):
+        model = torch.nn.Linear(4, 8)
+
+    model_handy = copy(model)
+    
+    register_colo_module(torch.nn.Linear, ColoLinear())
```
Comment (feifeibear) [inline tests/test_tensor/test_module_spec.py:45]: As _COLOSSAL_MODULES is a global dict. Why I have to register it in my code? Why not register all possible module automatically?

File: colossalai/tensor/module_handlers/colo_module_handler.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+        #          distspec.shard(xxxxx)
+        #    'row': ...
+        #    'col': ...
+        # }
+        self._allowed_patterns: Dict[ComputePattern, Dict[str, Dict[str, _DistSpec]]] = {}
+
+    def _register_shard_params(self, params: List[str]):
+        self._shard_params = params
+
+    def _register_allowed_patterns(self, compute_pattern: ComputePattern, dist_specs: Dict[str, _DistSpec], label='default'):
```
Comment (feifeibear) [inline colossalai/tensor/module_handlers/colo_module_handler.py:24]: use enum  class for label... str will bring typos.

File: tests/test_tensor/test_module_spec.py (hunk 7 above)
Comment (Wesley-Jzy) [reply tests/test_tensor/test_module_spec.py:45]: I prepare to register them in ColoInitContext. It's just for this unit test.

File: tests/test_tensor/test_module_spec.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    spec = TensorSpec(
+        distspec.shard(gpc.get_group(ParallelMode.PARALLEL_1D), [0], [gpc.get_world_size(ParallelMode.PARALLEL_1D)]),
+        ParallelAction(ComputePattern.TP1D))
+    with DistSpecManager.no_grad():
+        weight.set_spec(spec)
+        bias.set_spec(spec)
+
+
+def run_with_spec(spec_init_func):
+    with ColoInitContext(device=get_current_device()):
```
Comment (feifeibear) [inline tests/test_tensor/test_module_spec.py:40]: can you test a more complicated model, at least the simple_net?

File: colossalai/tensor/module_utils.py (hunk 2 above)
Comment (Wesley-Jzy) [reply colossalai/tensor/module_utils.py:8]: Done.

File: colossalai/tensor/module_utils.py (hunk 3 above)
Comment (Wesley-Jzy) [reply colossalai/tensor/module_utils.py:47]: Done.

File: colossalai/tensor/module_utils.py (hunk 4 above)
Comment (Wesley-Jzy) [reply colossalai/tensor/module_utils.py:31]: Done.

File: colossalai/tensor/module_utils.py (hunk 5 above)
Comment (Wesley-Jzy) [reply colossalai/tensor/module_utils.py:39]: Done.

File: colossalai/tensor/module_utils.py (hunk 6 above)
Comment (Wesley-Jzy) [reply colossalai/tensor/module_utils.py:79]: Done.

File: tests/test_tensor/test_module_spec.py (hunk 9 above)
Comment (Wesley-Jzy) [reply tests/test_tensor/test_module_spec.py:40]: I will test models in test_model today.

Comment (Wesley-Jzy) [reply tests/test_tensor/test_module_spec.py:40]: I will test models in test_model today.

**Comment by feifeibear** [comment]
What is the purpose to add module spec function?
I see you still need to set weight and bias spec? Is that make the param dist spec setting simple?

**Comment by Wesley-Jzy** [comment]
> What is the purpose to add module spec function? I see you still need to set weight and bias spec? Is that make the param dist spec setting simple?

It sets all params in a module and saves the allowed spec rules to avoid the unmatched spec in every param. Also, it can set the whole model with init_colo_module() after all ops are registered in init context.

File: tests/test_tensor/test_module_spec.py [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+
+def run_with_spec(label):
+    with ColoInitContext(device=get_current_device()):
+        model = torch.nn.Linear(4, 8)
+
+    model_handy = copy(model)
+    
+    register_colo_module(torch.nn.Linear, ColoLinear())
+    parallel_action = ParallelAction(ComputePattern.TP1D)
+    init_colo_module(model, parallel_action, recursive=True, label=label)
```
[the remark this answers is not in the document]
Comment (Wesley-Jzy) [reply tests/test_tensor/test_module_spec.py:30]: Add a simple_net test.

File: colossalai/tensor/module_utils.py [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+    if is_colo_module(module):
+        # for each param
+        # set DistSpec and ParallelAction
+        colo_module = get_colo_module(module)
+        if not colo_module.has_compute_pattern_with_label(compute_pattern, label=label):
+            raise NotImplementedError
+        for param_name, dist_spec in colo_module.get_dist_specs_with_label(compute_pattern, label=label).items():
+            if dist_spec is None:
+                continue
+            param = module.get_parameter(param_name)
```
[the remark this answers is not in the document]
Comment (Wesley-Jzy) [reply colossalai/tensor/module_utils.py:82]: Add a simple_net test.

**Review by feifeibear — APPROVED** [review]
